pupil_src/shared_modules/observable.py

```diff
@@ -187,29 +207,43 @@ def _get_wrapper_or_raise_if_not_exists(obj, method_name):
     observable_wrapper = getattr(obj, method_name)
     if not isinstance(observable_wrapper, _ObservableMethodWrapper):
         raise TypeError(
-            "Attribute {} of object {} is not observable. You never added an "
-            "observer to this!".format(method_name, obj)
+            f"Attribute '{method_name}' of object {obj} is not observable. You never "
+            "added an observer to this!"
         )
     return observable_wrapper
 
 
 class _ObservableMethodWrapper:
     def __init__(self, obj, method_name):
-        self._obj = obj
-        self._original_method = getattr(obj, method_name)
+        # the wrapper should not block garbage collection by reference counting for the
+        # observable object. Hence, we need to avoid strong references to the object,
+        # which would lead to cyclic references:
+        #   - The object is only referenced weakly
+        #   - The original wrapped method is only stored as an unbound method
+        self._obj_ref = weakref.ref(obj)
+        self._wrapped_func = getattr(obj.__class__, method_name)
         self._method_name = method_name
         self._observers = []
         self._was_removed = False
         self._patch_method_to_call_wrapper_instead()
 
     def _patch_method_to_call_wrapper_instead(self):
-        functools.update_wrapper(self, self._original_method)
-        setattr(self._obj, self._method_name, self)
+        functools.update_wrapper(self, self._wrapped_func)
+        # functools adds a reference to the wrapped method that we need to delete
+        # to avoid cyclic references
+        self.__wrapped__ = None
+        setattr(self._obj_ref(), self._method_name, self)
 
     def remove_wrapper(self):
-        setattr(self._obj, self._method_name, self._original_method)
+        setattr(obj, self._method_name, self._get_wrapped_bound_method())
```

Comment:
This shouldn't work, as `obj` is not defined here.
But anyways I think this method is not used anywhere? Can we just remove it? What was it used for? Is there anything missing?

Reply:
Thanks a lot for finding this, I thought this was covered by tests, so I didn't test it explicitly.

I fixed the method and added a test for it. And I ran a code coverage analysis and we cover now almost everything with the tests.

Still, we can discuss if we want to remove it. It's currently not used, but when I originally wrote `Observable` I thought it would be cleaner to make it possible to remove the wrapper and it's not a big chunk of code, so I don't see much harm in keeping it. Maybe @papr or @romanroibu have thoughts on this as well.

Reply:
Agree with @ckbaumann, if it's straight forward to implement and might be useful in the future, I don't think we should remove it. Especially since it's already there, and covered by tests. IMHO

Reply:
I agree, too. Resolving thread.